File: 03.Arboles/ej2.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Nodo {
    int valor;
    struct Nodo* izquierdo;
    struct Nodo* derecho;
    int altura;
} Nodo;

int obtenerAltura(Nodo* nodo) {
    if (nodo == NULL)
        return 0;
    return nodo->altura;
}

int maximo(int a, int b) {
    return (a > b) ? a : b;
}

Nodo* crearNodo(int valor) {
    Nodo* nuevo = (Nodo*)malloc(sizeof(Nodo));
    nuevo->valor = valor;
    nuevo->izquierdo = NULL;
    nuevo->derecho = NULL;
    nuevo->altura = 1;
    return nuevo;
}

int obtenerBalance(Nodo* nodo) {
    if (nodo == NULL)
        return 0;
    return obtenerAltura(nodo->izquierdo) - obtenerAltura(nodo->derecho);
}

Nodo* rotarDerecha(Nodo* y) {
    Nodo* x = y->izquierdo;
    Nodo* T2 = x->derecho;
    x->derecho = y;
    y->izquierdo = T2;
    y->altura = maximo(obtenerAltura(y->izquierdo), obtenerAltura(y->derecho)) + 1;
    x->altura = maximo(obtenerAltura(x->izquierdo), obtenerAltura(x->derecho)) + 1;
    return x;
}

Nodo* rotarIzquierda(Nodo* x) {
    Nodo* y = x->derecho;
    Nodo* T2 = y->izquierdo;
    y->izquierdo = x;
    x->derecho = T2;
    x->altura = maximo(obtenerAltura(x->izquierdo), obtenerAltura(x->derecho)) + 1;
    y->altura = maximo(obtenerAltura(y->izquierdo), obtenerAltura(y->derecho)) + 1;
    return y;
}

Nodo* insertar(Nodo* nodo, int valor) {
    if (nodo == NULL)
        return crearNodo(valor);
    if (valor < nodo->valor)
        nodo->izquierdo = insertar(nodo->izquierdo, valor);
    else if (valor > nodo->valor)
        nodo->derecho = insertar(nodo->derecho, valor);
    else
        return nodo;

    nodo->altura = 1 + maximo(obtenerAltura(nodo->izquierdo), obtenerAltura(nodo->derecho));
    int balance = obtenerBalance(nodo);

    if (balance > 1 && valor < nodo->izquierdo->valor)
        return rotarDerecha(nodo);
    if (balance < -1 && valor > nodo->derecho->valor)
        return rotarIzquierda(nodo);
    if (balance > 1 && valor > nodo->izquierdo->valor) {
        nodo->izquierdo = rotarIzquierda(nodo->izquierdo);
        return rotarDerecha(nodo);
    }
    if (balance < -1 && valor < nodo->derecho->valor) {
        nodo->derecho = rotarDerecha(nodo->derecho);
        return rotarIzquierda(nodo);
    }

    return nodo;
}

void recorridoEnOrden(Nodo* nodo) {
    if (nodo != NULL) {
        recorridoEnOrden(nodo->izquierdo);
        printf("%d ", nodo->valor);
        recorridoEnOrden(nodo->derecho);
    }
}

int main() {
    Nodo* raiz = NULL;
    int valores[] = {10, 20, 30, 40, 50, 25};  //tomo estos valores, podriamos sacarlos de un archivo, pero tendria que leer el archivo (faltaria el codigo)
    int n = sizeof(valores) / sizeof(valores[0]);

    for (int i = 0; i < n; i++) {
        raiz = insertar(raiz, valores[i]);
    }

    printf("Recorrido en orden del árbol AVL:\n");
    recorridoEnOrden(raiz);
    printf("\n");

    return 0;
}

File: 03.Arboles/ej2.py
class NodoAVL:
    def __init__(self, dato):
        self.dato = dato
        self.izq = None
        self.der = None
        self.altura = 1

class ArbolAVL:
    
    def obtener_altura(self, nodo):
        if not nodo:
            return 0
        return nodo.altura

    def obtener_balance(self, nodo):
        if not nodo:
            return 0
        return self.obtener_altura(nodo.izq) - self.obtener_altura(nodo.der)

    def rotar_derecha(self, y):
        x = y.izq
        T2 = x.der
        x.der = y
        y.izq = T2
        y.altura = 1 + max(self.obtener_altura(y.izq), self.obtener_altura(y.der))
        x.altura = 1 + max(self.obtener_altura(x.izq), self.obtener_altura(x.der))
        return x

    def rotar_izquierda(self, x):
        y = x.der
        T2 = y.izq
        y.izq = x
        x.der = T2
        x.altura = 1 + max(self.obtener_altura(x.izq), self.obtener_altura(x.der))
        y.altura = 1 + max(self.obtener_altura(y.izq), self.obtener_altura(y.der))
        return y

    def insertar(self, raiz, dato):
        if not raiz:
            return NodoAVL(dato)
        
        if dato < raiz.dato:
            raiz.izq = self.insertar(raiz.izq, dato)
        elif dato > raiz.dato:
            raiz.der = self.insertar(raiz.der, dato)
        else:
            return raiz

        raiz.altura = 1 + max(self.obtener_altura(raiz.izq), self.obtener_altura(raiz.der))
        balance = self.obtener_balance(raiz)

        # Caso Izquierda-Izquierda
        if balance > 1 and dato < raiz.izq.dato:
            return self.rotar_derecha(raiz)

        # Caso Derecha-Derecha
        if balance < -1 and dato > raiz.der.dato:
            return self.rotar_izquierda(raiz)

        # Caso Izquierda-Derecha
        if balance > 1 and dato > raiz.izq.dato:
            raiz.izq = self.rotar_izquierda(raiz.izq)
            return self.rotar_derecha(raiz)

        # Caso Derecha-Izquierda
        if balance < -1 and dato < raiz.der.dato:
            raiz.der = self.rotar_derecha(raiz.der)
            return self.rotar_izquierda(raiz)

        return raiz

    def mostrar_arbol(self, nodo, prefijo="", es_izq=True):
        if nodo:
            print(prefijo + ("└── " if es_izq else "├── ") + str(nodo.dato) + f" (Alt: {nodo.altura})")
            self.mostrar_arbol(nodo.izq, prefijo + ("    " if es_izq else "│   "), True)
            self.mostrar_arbol(nodo.der, prefijo + ("    " if es_izq else "│   "), False)

if __name__ == "__main__":
    avl = ArbolAVL()
    raiz = None
    datos_a_insertar = [10, 20, 30, 40, 50, 25]
    
    print("Insertando los valores:", datos_a_insertar)
    for valor in datos_a_insertar:
        raiz = avl.insertar(raiz, valor)
        
    print("\nEstructura final del Árbol AVL balanceado:")
    avl.mostrar_arbol(raiz)
